### PR TITLE
Implement Ngram scorer

### PR DESCRIPTION
**Pull request by qmpzzpmq**
hi, 

Here is Haoyu Tang from BIGO speech. I implement ngram with kenlm, and test it in aishell test since it works badly  in BPE.

It could improve only e2e model, but still not as good as RNNLM. it might since I didn't tuning the decoding parameters.

### CODE REVIEW

File: espnet/nets/pytorch_backend/lm/ngram.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import kenlm
+import torch
+
+from espnet.nets.scorer_interface import ScorerInterface
+from espnet.nets.scorer_interface import PartialScorerInterface
+
+from abc import ABC
+
+
+class Ngrambase(ABC):
```
Comment (ShigekiKarita) [inline espnet/nets/pytorch_backend/lm/ngram.py:10]: Can you add a small test for this class `test/test_ngram.py`? Maybe it can be just checking the output scores with toy inputs and arpa lm.

File: egs/aishell/asr1/path.sh [hunk 2]
```diff
@@ -2,7 +2,7 @@ MAIN_ROOT=$PWD/../../..
 KALDI_ROOT=$MAIN_ROOT/tools/kaldi
 
 [ -f $KALDI_ROOT/tools/env.sh ] && . $KALDI_ROOT/tools/env.sh
-export PATH=$PWD/utils/:$KALDI_ROOT/tools/openfst/bin:$KALDI_ROOT/tools/sctk/bin:$PWD:$PATH
+export PATH=$PWD/utils/:$KALDI_ROOT/tools/openfst/bin:$KALDI_ROOT/tools/sctk/bin:$PWD:$MAIN_ROOT/tools/kenlm/build/bin:$PATH
```
Comment (ShigekiKarita) [inline egs/aishell/asr1/path.sh:5]: This script should check if `tools/kenlm.done` exists. If it doesn't, show error messages like https://github.com/espnet/espnet/blob/292a15e51dca33e0339fa324062e5c4b5b8f4924/egs/fisher_callhome_spanish/st1/path.sh#L21-L27

File: egs/aishell/asr1/RESULTS.md [hunk 3]
```diff
@@ -36,3 +36,21 @@ exp/train_sp_pytorch_no_patience/decode_test_beam20_emodel.acc.best_p0.0_len0.0-
 |    SPKR       |     # Snt         # Wrd     |     Corr           Sub            Del           Ins            Err         S.Err     |
 |    Sum/Avg    |     7176         104765     |     92.2           7.6            0.2           0.2            8.0          50.2     |
 ```
+
+# Ngram related 
+   - there is no RNN not ngram
```
Comment (ShigekiKarita) [inline egs/aishell/asr1/RESULTS.md:41]: This result looks promising. Have you ever tried ngram-RNNLM interpolation (joint decoding)?

Comment (sw005320) [reply egs/aishell/asr1/RESULTS.md:41]: That's my recommendation, too, and @qmpzzpmq is trying now.

Comment (qmpzzpmq) [reply egs/aishell/asr1/RESULTS.md:41]: I tried, but there is no result better than currently one.
I guess it takes time to tuning the decoding weight parameters.

**Comment by qmpzzpmq** [comment]
@ShigekiKarita 
Hi, I am not I toally understand your means about test_ngram.py, but I still tried my best to finised to do it. This program to check is the score same with I got

File: test/test_ngram.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+from math.isclose
+
+from espnet.nets.pytorch_backend.lm.ngram import NgramFullScorer
+from espnet.nets.pytorch_backend.lm.ngram import NgramPartScorer
+
+test_sens = ["I like apple", "you love coffee", "I love you", "I love coffee"]
+@pytest.mark.parameterize(test_sens)
+def test_ngram_build(test_sens):
+    os.system(f"lmplz --discount_fallback -o 3 {'\n'.join(test_sens)} >tmp.arpa")
+    os.system("build_binary tmp.arpa tmp.bin")
```
Comment (ShigekiKarita) [inline test/test_ngram.py:13]: This test is OK. However, I prefer storing `test/ngram.arpa` and `test/ngram.bin` rather than generating them in test because this test requires `kenlm` installation.

File: tools/Makefile [hunk 5]
```diff
@@ -25,7 +25,7 @@ endif
 
 all: kaldi.done python check_install
 
-python: venv warp-ctc.done warp-transducer.done chainer_ctc.done
+python: venv warp-ctc.done warp-transducer.done chainer_ctc.done kenlm.done
```
Comment (ShigekiKarita) [inline tools/Makefile:28]: https://github.com/espnet/espnet/pull/1946/checks?check_run_id=694960347 
As seen in this CI error, I concern that installing kenlm seems so tough. This is because it requires Boost. I think it should be optional. Consider `pip install https://github.com/kpu/kenlm/archive/master.zip` for python installation. https://github.com/kpu/kenlm

Comment (ShigekiKarita) [reply tools/Makefile:28]: I think the best way to install kenlm python binding is adding the github URL to `setup.py` like https://github.com/espnet/espnet/blob/66f0a1263e05848482377ad6c515b03ccb20d594/setup.py#L48-L49

Comment (qmpzzpmq) [reply tools/Makefile:28]: but it needs "lmplz" and "build_binary"

File: test/test_ngram.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import pytest
+import os
+import kenlm
```
Comment (ShigekiKarita) [inline test/test_ngram.py:3]: If you think even kenlm python binding is difficult to install by default, you can skip this test by `kenlm = pytest.importorskip("kenlm")`  https://docs.pytest.org/en/latest/skipping.html#skipping-on-a-missing-import-dependency

File: test/test_ngram.py (hunk 4 above)
Comment (qmpzzpmq) [reply test/test_ngram.py:13]: I thinks store .arpa file only is enough

File: espnet/asr/pytorch_backend/recog.py [hunk 7]
```diff
@@ -65,16 +66,22 @@ def recog_v2(args):
     else:
         lm = None
 
+    if args.ngram_model:
+        ngram = NgramFullScorer(args.ngram_model, train_args.char_list)
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/recog.py:70]: Maybe you should import `kenlm` and its scorer module locally if the python binding is optional.

Comment (qmpzzpmq) [reply espnet/asr/pytorch_backend/recog.py:70]: maybe import it in init function of Ngrambase

Comment (qmpzzpmq) [reply espnet/asr/pytorch_backend/recog.py:69]: done

**Comment by qmpzzpmq** [comment]
@ShigekiKarita is everything alright?

File: tools/Makefile [hunk 8]
```diff
@@ -78,6 +78,13 @@ chainer_ctc.done: espnet.done
 	. venv/bin/activate && ./install_chainer_ctc.sh
 	touch chainer_ctc.done
 
+kenlm.done: espnet.done
+	rm -rf kenlm
+	git clone https://github.com/kpu/kenlm.git
+	mkdir -p kenlm/build && cd kenlm/build && cmake .. && make -j4;
```
Comment (b-flo) [inline tools/Makefile:84]:
```suggestion
	mkdir -p kenlm/build && cd kenlm/build && cmake .. && $(MAKE);
```

File: tools/Makefile (hunk 5 above)
Comment (kamo-naoyuki) [reply tools/Makefile:28]: @ShigekiKarita Now we can't use direct installation from url in setup.py, `@git+https://`, because pypi server doesn't allow it. ( I registered espnet to pypi recently )

@qmpzzpmq 
- Change the installation for kenlm to an option.
- Remove the dependencies on "lmplz" and "build_binary" on the test. It can be avoided by using stored arpa and bin file.
- Install kenlm by `pip install https://github.com/kpu/kenlm/archive/master.zip` for the unittest: Here https://github.com/espnet/espnet/blob/develop/ci/install.sh

**Comment by kamo-naoyuki** [comment]
Could you fix the CI error?

File: tools/Makefile [hunk 9]
```diff
@@ -78,6 +78,13 @@ chainer_ctc.done: espnet.done
 	. venv/bin/activate && ./install_chainer_ctc.sh
 	touch chainer_ctc.done
 
+kenlm.done: espnet.done
+	rm -rf kenlm
+	git clone https://github.com/kpu/kenlm.git
```
Comment (kamo-naoyuki) [inline tools/Makefile:83]: Create install_kenlm.sh as same as the other instead of writing it in the Makefile.

File: egs/aishell/asr1/test.txt [hunk 10]
```diff
@@ -0,0 +1,4 @@
+I like apple
+you love coffee
```
Comment (kamo-naoyuki) [inline egs/aishell/asr1/test.txt:2]: What is this file?

File: egs/tedlium2/asr1/test.sh [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+. ./path.sh
+lmplz
```
Comment (kamo-naoyuki) [inline egs/tedlium2/asr1/test.sh:3]: What is this file?

File: egs/aishell/asr1/test.arpa [hunk 12]
```diff
@@ -0,0 +1,42 @@
+\data\
+ngram 1=9
```
Comment (kamo-naoyuki) [inline egs/aishell/asr1/test.arpa:2]: Move this file to test/

**Comment by qmpzzpmq** [comment]
@kamo-naoyuki done, I am not family with flake8, so it takes times, I will configure my vsc next time.

File: espnet/nets/pytorch_backend/lm/ngram.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+    def __init__(self, ngram_model, token_list):
+        """Initialize Ngrambase.
+
+        Args:
+            ngram_model: ngram model path
+            token_list: token list from dict or model.json
+
+        """
+        self.chardict = [x if x != "<eos>" else "</s>" for x in token_list]
```
Comment (kamo-naoyuki) [inline espnet/nets/pytorch_backend/lm/ngram.py:23]: I have a question.
In espnet, `<eos>` represents `<s>` and </s> both, but you only care about `</s>`. Is this no problem?

File: tools/Makefile [hunk 14]
```diff
@@ -27,7 +27,7 @@ all: kaldi.done python check_install
 
 python: venv warp-ctc.done warp-transducer.done chainer_ctc.done
 
-extra: nkf.done moses.done mwerSegmenter.done pesq
+extra: nkf.done moses.done mwerSegmenter.done pesq kenlm.done
```
Comment (kamo-naoyuki) [inline tools/Makefile:31]: Would you add the following block?

```makefile
kenlm.done:
	. venv/bin/activate && ./install_kenlm.sh
	touch kenlm.done
```

**Comment by kamo-naoyuki** [comment]
Did you delete test_ngram.py? Could you restore it?

File: espnet/nets/pytorch_backend/lm/ngram.py (hunk 13 above)
Comment (qmpzzpmq) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: I have descripted in the program since the https://github.com/qmpzzpmq/espnet/blob/8bf8db9e855ceef181e4fbce4a610724da283c3d/espnet/nets/pytorch_backend/lm/ngram.py#L31

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: `<eos>` represents both `<s>` and `</s>` (=same label is used in espnet) and the hypothesis always start with `<eos>`. 
Am I wrong? I didn't read the detail of the code.

File: test/test_ngram.py [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+from math import isclose
+
+kenlm = pytest.importorskip("kenlm")
+
+test_sens = ["I like apple", "you love coffee"]
+
+
+@pytest.mark.parametrize("test_sens", [test_sens])
+def test_ngram_build(test_sens):
+    lm = kenlm.LanguageModel("test/test.arpa")
```
Comment (kamo-naoyuki) [inline test/test_ngram.py:12]: `test/test.arpa` supposes the executing directory level.

The following file is a good example.
https://github.com/espnet/espnet/blob/master/test/test_sentencepiece.py

File: espnet/nets/pytorch_backend/lm/ngram.py (hunk 13 above)
Comment (qmpzzpmq) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: please read this one:
https://github.com/kpu/kenlm/blob/87e85e66c99ceff1fab2500a7c60c01da7315eec/python/example.py#L34-L35

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: The examples just mentions the case without `<s>`, but our case, it starts with `<s>`.

File: test/test_ngram.py (hunk 15 above)
Comment (qmpzzpmq) [reply test/test_ngram.py]: done

File: espnet/nets/pytorch_backend/lm/ngram.py (hunk 13 above)
Comment (qmpzzpmq) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: Yes, I recheck it. I correct it.

Comment (kamo-naoyuki) [reply espnet/nets/pytorch_backend/lm/ngram.py:23]: Thank you, lgtm

File: egs/tedlium2/asr1/conf/tuning/decode_ngram.yaml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+ngram-weight: 1.0
+beam-size: 20
```
Comment (kamo-naoyuki) [inline egs/tedlium2/asr1/conf/tuning/decode_ngram.yaml:2]: run.sh for tedliums2 doesn't exist?
Is this a garbage file?

File: egs/tedlium2/asr1/path.sh [hunk 17]
```diff
@@ -2,10 +2,11 @@ MAIN_ROOT=$PWD/../../..
 KALDI_ROOT=$MAIN_ROOT/tools/kaldi
```
Comment (kamo-naoyuki) [inline egs/tedlium2/asr1/path.sh:3]: ditto

**Comment by kamo-naoyuki** [comment]
@ShigekiKarita What do you think about the path of ngram scorer?  I think espnet/nets/scorers is better.

File: egs/aishell/asr1/run.sh [hunk 18]
```diff
@@ -153,7 +157,7 @@ if [ -z ${lmtag} ]; then
 fi
 lmexpname=train_rnnlm_${backend}_${lmtag}
 lmexpdir=exp/${lmexpname}
-mkdir -p ${lmexpdir}
+# [ ! -d ${lmexpdir} ] && mkdir -p ${lmexpdir}
```
Comment (sw005320) [inline egs/aishell/asr1/run.sh:160]: I'm not sure your intension but you can just use the original one.
```suggestion
mkdir -p ${lmexpdir}
```

File: egs/aishell/asr1/run.sh [hunk 19]
```diff
@@ -178,6 +182,26 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
         --dict ${dict}
 fi
 
+ngramexpname=train_ngram
```
Comment (sw005320) [inline egs/aishell/asr1/run.sh:185]: Can you include this ngram LM training in stage 3?
I want to keep the role of the stage.

**Comment by sw005320** [comment]
> @ShigekiKarita What do you think about the path of ngram scorer? I think espnet/nets/scorers is better.

I agree.

File: egs/aishell/asr1/run.sh (hunk 19 above)
Comment (qmpzzpmq) [reply egs/aishell/asr1/run.sh]: done

File: egs/tedlium2/asr1/path.sh (hunk 17 above)
Comment (qmpzzpmq) [reply egs/tedlium2/asr1/path.sh]: done

File: egs/tedlium2/asr1/conf/tuning/decode_ngram.yaml (hunk 16 above)
Comment (qmpzzpmq) [reply egs/tedlium2/asr1/conf/tuning/decode_ngram.yaml]: done

File: egs/aishell/asr1/run.sh [hunk 20]
```diff
@@ -176,8 +187,20 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
         --valid-label ${lmdatadir}/valid.txt \
         --resume ${lm_resume} \
         --dict ${dict}
+    
+    echo "stage 4: Ngram Preparation"
```
Comment (sw005320) [inline egs/aishell/asr1/run.sh:191]:
```suggestion
    echo "stage 3: Ngram Preparation"
```

**Comment by sw005320** [comment]
Very cool!
One more request.
Since this is a very cool feature, please add this n-gram extension in main `REAME.md` (https://github.com/espnet/espnet#key-features)

**Comment by qmpzzpmq** [comment]
> Very cool!
> One more request.
> Since this is a very cool feature, please add this n-gram extension in main `REAME.md` (https://github.com/espnet/espnet#key-features)
@sw005320 done

**Comment by sw005320** [comment]
I think this PR is almost ready. I just want to make sure that @ShigekiKarita would agree with putting this in `espnet/nets/scorers`.

**Comment by b-flo** [comment]
Great work!
On another note, for your next PR you should try to group your commits a bit more (e.g.: consecutive commits with same target/purpose but small changes in each). I wanted to follow the changes but I had to unsubscribe to the PR because an email is sent for each commit you did (50+ in a few days!)...

**Comment by qmpzzpmq** [comment]
@b-flo 
yes, I will. since it is my first time to add a total new feature to espnet. I am not family with your test process, but now I know something about it.
 And I am work on MoCha recently.
